Document SHOW FUNCTIONS and SHOW PROCEDURES

Hunterness opened the pull request:
Depends on ~~https://github.com/neo-technology/neo4j/pull/9840 and https://github.com/neo-technology/neo4j/pull/9908~~

Companion to https://github.com/neo-technology/neo4j-manual-modeling/pull/1982

(I have NOT built and looked at this)

Hunterness committed:
Document SHOW FUNCTIONS and SHOW PROCEDURES

Hunterness committed:
Update to not use the deprecated procedures

mnd999 commented:
Getting a failure in the docs build:
```
Exception in thread "main" org.neo4j.graphdb.QueryExecutionException: Unsupported administration command: SHOW ALL PRIVILEGES YIELD * WHERE action='execute' AND segment STARTS WITH $seg RETURN access, segment, collect(role) as roles
	at org.neo4j.kernel.impl.query.QueryExecutionKernelException.asUserException(QueryExecutionKernelException.java:35)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:262)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:243)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.showcommands.ShowProcFuncCommandHelper$.getPrivileges(ShowProcFuncCommandHelper.scala:107)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.showcommands.ShowProceduresCommand.originalNameRows(ShowProceduresCommand.scala:46)
	at org.neo4j.cypher.internal.runtime.interpreted.commands.showcommands.Command.rows(Command.scala:32)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.CommandPipe.internalCreateResults(CommandPipe.scala:28)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.Pipe.createResults(Pipe.scala:43)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.Pipe.createResults$(Pipe.scala:40)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.CommandPipe.createResults(CommandPipe.scala:26)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.PipeWithSource.createResults(Pipe.scala:67)
	at org.neo4j.cypher.internal.runtime.interpreted.PipeExecutionResult.request(PipeExecutionResult.scala:63)
	at org.neo4j.cypher.internal.result.StandardInternalExecutionResult.request(StandardInternalExecutionResult.scala:90)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.request(ClosingExecutionResult.scala:144)
	at org.neo4j.cypher.internal.javacompat.ResultSubscriber.fetchResults(ResultSubscriber.java:355)
	at org.neo4j.cypher.internal.javacompat.ResultSubscriber.nextFromSubscriber(ResultSubscriber.java:323)
	at org.neo4j.cypher.internal.javacompat.ResultSubscriber.fetchNextOrNull(ResultSubscriber.java:294)
	at org.neo4j.cypher.internal.javacompat.ResultSubscriber.fetchNextOrNull(ResultSubscriber.java:58)
	at org.neo4j.internal.helpers.collection.PrefetchingIterator.peek(PrefetchingIterator.java:60)
	at org.neo4j.internal.helpers.collection.PrefetchingIterator.hasNext(PrefetchingIterator.java:46)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:132)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.neo4j.doc.ProcedureReferenceGenerator.parseResult(ProcedureReferenceGenerator.java:116)
	at org.neo4j.doc.ProcedureReferenceGenerator.procedures(ProcedureReferenceGenerator.java:108)
	at org.neo4j.doc.ProcedureReferenceGenerator.communityEditionProcedures(ProcedureReferenceGenerator.java:89)
	at org.neo4j.doc.ProcedureReferenceGenerator.document(ProcedureReferenceGenerator.java:63)
	at org.neo4j.doc.GenerateProcedureReference.main(GenerateProcedureReference.java:62)
```

mnd999 commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/QueryPlanTest.scala:
```diff
@@ -379,6 +379,33 @@ class QueryPlanTest extends DocumentingTestBase with SoftReset {
   }
 
+  @Test def showFunctions() {profileQuery(
+      title = "Listing functions",
```
We use "List indexes" and "List constraints" but "Listing" here - we should be consistent.

Hunterness replied:
But we already have `List functions` which is functions of category `List`... (I started by adding the functions to the same chapter as that hence not wanting to use the same name). 

I also don't want to risk confusing these with lists in cypher...

But I agree that we should be consistent... Not sure if it is better to update indexes/constraints but I guess the admin commands also use list (or maybe listing?)

Hunterness replied:
Since all administration `SHOW` commands have `Listing` I'll update indexes/constraints to use that as well for consistency

mnd999 replied:
I think you missed one:
![image](https://user-images.githubusercontent.com/1756429/117945757-8b1c5e80-b306-11eb-8da4-92b5131b4158.png)

QueryPlanTest, I think.

mnd999 commented on cypher/cypher-docs/src/docs/dev/clause.asciidoc:
```diff
@@ -239,6 +240,22 @@ m|--- <<query-using-periodic-commit-hint, USING PERIODIC COMMIT>>
 
 
+[[header-listing-clauses]]
+**Listing clauses**
```
I don't really like "Listing Clauses", but I'm struggling to think of anything better. Maybe we can go more specific - "Listing Procedures and Functions", but that doesn't scale well. Metadata, and catalogue are too vague. Maybe DBMS [query] clauses? Perhaps docs have a better suggestion.

Hunterness replied:
> I don't really like "Listing Clauses", but I'm struggling to think of anything better.

That was my exact problem

davidoliverSP2 replied:
@Hunterness and I discussed this, and thought that **Listing functions and procedures** is the best bet for now

Hunterness replied:
Can add the reasoning:
1. We are not sure what is coming down the pipeline so we might as well keep it explicit now and update it to be more inclusive when we know what to add.
2. The one thing we do know is coming are transaction and query commands, which will be both listing and terminating. This means they will most likely get their own section anyway (or the section needs to be updated to `management` or something to include terminate as well and then we might as well be descriptive now).

mnd999 commented on cypher/cypher-docs/src/docs/dev/ql/administration/constraints/index.asciidoc:
```diff
@@ -75,7 +75,7 @@ Using the keyword `IF NOT EXISTS` makes the command idempotent, and no error wil
 == Syntax
 
-.Syntax for managing indexes
+.Syntax for managing constraints
```
This doesn't seem related, is it intentional?

Hunterness replied:
I copied this tale to use as a base for the procedure/function ones and realized the table for `CONSTRAINTS` said `INDEXES` so I updated that while I noticed it

So it is intentional and technically unrelated

mnd999 commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ShowFunctionsTest.scala:
```diff
@@ -0,0 +1,91 @@
+        query("SHOW FUNCTIONS", ResultAssertions(p => {
+          p.columns should contain theSameElementsAs Array("name", "category", "description")
+        })) {
+          resultTable()
```
This table is very large - probably we should restrict it with LIMIT, or omit it entirely.

Hunterness replied:
It it very large but I'm unsure if I want to not display it at all. And I wanted to have the `SHOW` command without any extra clauses so I can't have limit

mnd999 commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ShowFunctionsTest.scala:
```diff
@@ -0,0 +1,91 @@
+          p.columnAs[List[String]]("rolesExecution").foreach(_ should be(null))
+          p.columnAs[List[String]]("rolesBoostedExecution").foreach(_ should be(null))
+        })) {
+          resultTable()
```
Again, this is huge, please restrict the output, and lots of columns make it not-so-readable:
![image](https://user-images.githubusercontent.com/1756429/117287372-2d00fe80-ae62-11eb-92a9-46e3ff5b0e3f.png)

Hunterness replied:
this I could add limit to (and maybe yield specific instead of `*`)

mnd999 commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ShowFunctionsTest.scala:
```diff
@@ -0,0 +1,91 @@
+        query("SHOW FUNCTIONS EXECUTABLE BY jake", ResultAssertions(p => {
+          p.columns should contain theSameElementsAs Array("name", "category", "description")
+        })) {
+          resultTable()
```
Again, this is a lot

mnd999 commented on cypher/cypher-docs/src/docs/dev/ql/listing/listing-procedures-columns.adoc:
```diff
@@ -0,0 +1,67 @@
+| `+`
+
+| `mode`
+| The procedure mode, for example `READ` or `WRITE`.
```
Is there a reference for procedure modes we can link to from here?

Hunterness replied:
maybe, but probably not, can see if I can find anything

Hunterness replied:
They are described in the java reference in the section about writing [user-defined procedures](https://neo4j.com/docs/java-reference/current/extending-neo4j/procedures-and-functions/procedures/#writing-procedure), but I don't find anything else when searching the docs

mnd999 commented on cypher/cypher-docs/src/docs/dev/ql/listing/listing-procedures-columns.adoc:
```diff
@@ -0,0 +1,67 @@
+| `+`
+
+| `admin`
+| If it is or isn't an admin procedure.
```
```suggestion
`true` if this procedure is an admin procedure
```

mnd999 commented on cypher/cypher-docs/src/docs/dev/ql/listing/listing-procedures-columns.adoc:
```diff
@@ -0,0 +1,67 @@
+| `+`
+
+| `rolesExecution`
+| List of roles who can execute this procedure.
```
```suggestion
| List of roles permitted to execute this procedure.
```

mnd999 commented on cypher/cypher-docs/src/docs/dev/ql/listing/listing-procedures-columns.adoc:
```diff
@@ -0,0 +1,67 @@
+| `+`
+
+| `rolesBoostedExecution`
+| List of roles who will use boosted mode when executing this procedure.
```
```suggestion
| List of roles permitted to use boosted mode when executing this procedure.
```

mnd999 commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ShowProceduresTest.scala:
```diff
@@ -0,0 +1,88 @@
+        query("SHOW PROCEDURES", ResultAssertions(p => {
+          p.columns should contain theSameElementsAs Array("name", "description", "mode", "worksOnSystem")
+        })) {
+          resultTable()
```
Again, this is 88 rows, please limit somehow

mnd999 commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ShowProceduresTest.scala:
```diff
@@ -0,0 +1,88 @@
+          p.columns should contain theSameElementsAs Array("name", "admin")
+          p.columnAs[Boolean]("admin").foreach(_ should be(true))
+        })) {
+          resultTable()
```
Maybe swap the columns here would be nicer:
![image](https://user-images.githubusercontent.com/1756429/117288933-f1673400-ae63-11eb-9335-9ddb5c91a490.png)

Hunterness replied:
That is a problem all over and comes from me not wanting every example to have a return clause and therefore getting the columns in alphabetical order. This one I think could have it though

mnd999 commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ShowProceduresTest.scala:
```diff
@@ -0,0 +1,88 @@
+          p.columnAs[List[String]]("rolesExecution").foreach(_ should be(null))
+          p.columnAs[List[String]]("rolesBoostedExecution").foreach(_ should be(null))
+        })) {
+          resultTable()
```
Again, many narrow columns are hard to read and 57 rows is a lot:
![image](https://user-images.githubusercontent.com/1756429/117289114-22dfff80-ae64-11eb-97d4-745a3eab613a.png)

Hunterness replied:
I know, all of these are horrible result tables.

But I don't want to get rid of all tables and I want to have different types of queries and not just have yield and return on all of them so I'm not sure how to fix this...

I could maybe try to make some custom resultTable methods that just prints the first 10 rows and reorder the columns to the order we want. can't really do anything about the amount of columns when using `yield *` though

mnd999 replied:
I think a shorter result table would be generally useful, so yeah that sounds good. Maybe it could also limit the number of columns and just show an empty column at the end with '...' as the header row. Not sure how easy that is to do though.

Hunterness replied:
Any thoughts as to how many rows would be a good cut-off?

mnd999 commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ShowProceduresTest.scala:
```diff
@@ -0,0 +1,88 @@
+        query("SHOW PROCEDURES EXECUTABLE BY jake", ResultAssertions(p => {
+          p.columns should contain theSameElementsAs Array("name", "description", "mode", "worksOnSystem")
+        })) {
+          resultTable()
```
Again, 57 rows please restrict.

mnd999 commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ShowProceduresTest.scala:
```diff
@@ -0,0 +1,88 @@
+        }
+        p("Notice that the two roles columns are empty due to missing the <<administration-security-administration-dbms-privileges-role-management, SHOW ROLE>> privilege.")
+        logout()
+        p("The second versions is to filter for a specific user:")
```
```suggestion
        p("The second version filters the list to only contain procedures executable by a specific user:")
```

mnd999 commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ShowProceduresTest.scala:
```diff
@@ -0,0 +1,88 @@
+        )
+        database("neo4j")
+        p(
+          """The listed procedures can also be filtered on whether a user can execute them.
```
```suggestion
          """The listed procedures can also be filtered by whether a user can execute them.
```

mnd999 commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/ShowFunctionsTest.scala:
```diff
@@ -0,0 +1,91 @@
+      section("Listing all functions") {
+        p(
+          """To list all available functions with the default output columns, the `SHOW FUNCTIONS` command can be used.
+            |If all columns are wanted, use `SHOW FUNCTIONS YIELD *`.""".stripMargin)
```
```suggestion
            |If all columns are required, use `SHOW FUNCTIONS YIELD *`.""".stripMargin)
```

mnd999 commented on cypher/cypher-docs/src/docs/dev/ql/listing/listing-functions-columns.adoc:
```diff
@@ -0,0 +1,62 @@
+| `+`
+
+| `rolesExecution`
+| List of roles who can execute this function.
```
```suggestion
| List of roles permitted to execute this function.
```

mnd999 commented on cypher/cypher-docs/src/docs/dev/ql/listing/listing-functions-columns.adoc:
```diff
@@ -0,0 +1,62 @@
+| `+`
+
+| `rolesBoostedExecution`
+| List of roles who will use boosted mode when executing this function.
```
```suggestion
| List of roles permitted to use boosted mode when executing this function.
```

mnd999 commented on cypher/cypher-docs/src/docs/dev/ql/listing/listing-functions-columns.adoc:
```diff
@@ -0,0 +1,62 @@
+| `+`
+
+| `aggregating`
+| If the function is aggregating or not.
```
```suggestion
| Whether the function is aggregating or not.
```

Hunterness committed:
Address general review comments

Hunterness committed:
Restrict the printout of procedures and functions

There are so many built-in functions and procedures that the tables was way to big
now we limit the amount of rows and columns printed

It will also print the columns in the given order

Hunterness committed:
Change the stripmargin

Have noticed parts of the docs have it this way instead

mnd999 commented:
As discussed, one new observation, the Listings Clauses seem to be inlined on the Clauses page which leads to the menu on the RHS:

![image](https://user-images.githubusercontent.com/1756429/117944996-c9654e00-b305-11eb-8801-f4934370fa4e.png)

Hunterness committed:
Missed list -> listing

Hunterness commented:
A bit more info around the listing clauses and content list: The other clauses (like `MATCH`) gets put on separate pages and linked in the left side content list. The listing clauses gets put as new sections on the same page as `Clauses` instead and therefore in the right hand side contents.

For docs team reviewer: Is there a way to get `SHOW PROCEDURE` and `SHOW FUNCTIONS` to become separate pages as well and behaving the same way as the other clauses?

davidoliverSP2 commented:
> For docs team reviewer: Is there a way to get `SHOW PROCEDURE` and `SHOW FUNCTIONS` to become separate pages as well and behaving the same way as the other clauses?

@Hunterness I'll have a go, and I'll push to the PR if I'm able to!

davidoliverSP2 commented:
> As discussed, one new observation, the Listings Clauses seem to be inlined on the Clauses page which leads to the menu on the RHS:
> 
> ![image](https://user-images.githubusercontent.com/1756429/117944996-c9654e00-b305-11eb-8801-f4934370fa4e.png)

Fixed in https://github.com/neo-technology/neo4j-manual-modeling/pull/1982/commits/1c2416b17564ea8a5808d73d1de6567a74aeb438

davidoliverSP2 committed:
rename functions and procedures pages to match other files

davidoliverSP2 committed:
rename sub-section

Hunterness commented on cypher/cypher-docs/src/docs/dev/clause.asciidoc:
```diff
@@ -240,8 +240,8 @@ m|--- <<query-using-periodic-commit-hint, USING PERIODIC COMMIT>>
 
-[[header-listing-clauses]]
-**Listing clauses**
+[[header-listing-procs-functions]]
```
We are referencing to this in one more place

Hunterness commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/ShowProcFuncTest.scala:
```diff
@@ -0,0 +1,59 @@
+class ShowProcFuncTest extends RefcardTest with QueryStatisticsTestSupport {
+  val graphDescription = List()
+  val title = "SHOW FUNCTIONS and PROCEDURES"
+  override val linkId = "clauses/#header-listing-clauses"
```
```suggestion
  override val linkId = "clauses/#header-listing-procs-functions"
```

Hunterness committed:
Update missed link change

davidoliverSP2 reviewed (approved):
Looks good, thanks @Hunterness 

I just need to verify the refcard changes in the corresponding https://github.com/neo-technology/neo4j-manual-modeling/pull/1982
